sudoku_board.cc

```diff
@@ -1,123 +1,181 @@
+                                                         6, 7, 8, 9};
+const int SudokuBoard::EMPTY_CELL_SIGN = 0;
+const size_t SudokuBoard::CUBE_SIZE = 3;
+const std::string SudokuBoard::OPTION_INVALID_ERR_MSG =
+    "The givven value doesn't appear in the possible options.";
+const std::string SudokuBoard::SIZE_INCOMPATIBALE_ERR_MSG =
+    "The other board should be in the same size as this board.";
+
+// constructors
+
+SudokuBoard::SudokuBoard(size_t size)
+    : m_size(size),
+      m_board_values(size, std::vector<int>(size, EMPTY_CELL_SIGN)),
+      m_possible_options(
+          size, std::vector(size, std::set<int>(POSSIBLE_CELL_VALUES))) {}
+
+SudokuBoard::SudokuBoard(std::vector<std::vector<int>> board_values)
+    : m_size(board_values.size()), m_board_values(std::move(board_values)),
+      m_possible_options(
+          m_size, std::vector(m_size, std::set<int>(POSSIBLE_CELL_VALUES))) {
+  update_possible_options();
 }
 
-SudokuBoard::SudokuBoard(int size, std::vector<std::vector<int>> board_values){
-    m_size = size;
-    m_board_values = board_values; //question: when to use a copy constructor?
-    m_possible_options = std::vector(size, std::vector(size, std::set<int>(POSSIBLE_CELL_VALUES)));
-    update_possible_options();
+// getters
+
+size_t SudokuBoard::get_size() const { return m_size; }
+
+const std::vector<std::vector<int>> &SudokuBoard::get_board_values() const {
+  return m_board_values;
 }
 
-int SudokuBoard::get_size(){
-    return m_size;
+const std::vector<std::vector<std::set<int>>> &
+SudokuBoard::get_possible_options() const {
+  return m_possible_options;
 }
 
-// question: do I need to return a copy and not the original one?
-std::vector<std::vector<int>> SudokuBoard::get_board_values(){
-    return m_board_values;
+const std::set<int> &SudokuBoard::get_cell_options(size_t r, size_t c) const {
+  return m_possible_options[r][c];
 }
 
-std::vector<std::vector<std::set<int>>> SudokuBoard::get_possible_options(){
-    return m_possible_options;
+size_t SudokuBoard::get_missing_cells_amount() const {
+  size_t missing_cells_amount = 0;
+  for (size_t i = 0; i < m_size; i++) {
+    missing_cells_amount += count(m_board_values[i].begin(),
+                                  m_board_values[i].end(), EMPTY_CELL_SIGN);
+  }
+  return missing_cells_amount;
 }
 
-STATUS_OPTION SudokuBoard::get_status(){
-    bool is_win = true;
-    for (int i = 0; i < m_size; i++){
-        for (int j = 0; j < m_size; j++){
-            STATUS_OPTION cell_status = get_cell_status(i, j);
-            if (cell_status == LOSE){
-                return LOSE;
-            }
-            else if (cell_status == PROG){
-                is_win = false;
-            }
-        }
-    }
-    if (is_win){
-        return WIN;
-    }
-    return PROG;
+// setters
+
+void SudokuBoard::set_cell_options(size_t r, size_t c,
+                                   std::set<int> cell_options) {
+  m_possible_options[r][c] = cell_options;
```

Comment:
```suggestion
  m_possible_options[r][c] = std::move(cell_options);
```

You avoid a copy this way